crates/volta-core/src/hook/mod.rs

```diff
@@ -68,9 +72,79 @@ pub struct ToolHooks<D: Distro> {
     pub phantom: PhantomData<D>,
 }
 
+impl<D: Distro> ToolHooks<D> {
+    /// Creates a merged struct, with "right" having precedence over "left".
+    fn merge(left: Self, right: Self) -> Self {
+        Self {
+            distro: if let Some(_) = right.distro {
+                right.distro
+            } else {
+                left.distro
+            },
+            latest: if let Some(_) = right.latest {
+                right.latest
+            } else {
+                left.latest
+            },
+            index: if let Some(_) = right.index {
+                right.index
+            } else {
+                left.index
+            },
+            phantom: PhantomData,
+        }
+    }
+}
+
 impl HookConfig {
-    /// Returns the current hooks, loaded from the filesystem.
+    /// Returns the current hooks, which are a merge between the user hooks and
+    /// the project hooks (if any).
     fn current() -> Fallible<Self> {
+        let maybe_project_config = Self::for_current_dir()?;
+        let user_config = Self::for_user()?;
+
+        Ok(if let Some(project_config) = maybe_project_config {
+            Self::merge(user_config, project_config)
+        } else {
+            user_config
+        })
+    }
+
+    /// Returns the per-project hooks for the current directory.
+    fn for_current_dir() -> Fallible<Option<Self>> {
+        Self::for_dir(&env::current_dir().with_context(|_| ErrorDetails::CurrentDirError)?)
+    }
+
+    /// Returns the per-project hooks for the specified directory.  If the
+    /// specified directory is not itself a project, its ancestors will be
+    /// searched.
+    fn for_dir(base_dir: &Path) -> Fallible<Option<Self>> {
+        let mut dir = base_dir.clone();
+        while !is_project_root(dir) {
```

Comment:
Since we're using the `is_project_root` as the differentiator, it would be nice to abstract this logic into a `find_project_root` method that takes `base_dir: &Path` in and returns an `Option<&Path>`. Then both this and the `project.rs` code can use that method, instead of duplicating the code for the ancestor searching.

Reply:
👍, will do.